File: frontend/lib/slices/bus-closing-voucher.ts
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../store';

// Updated BusClosingVoucher type based on your new requirements
export type BusClosingVoucher = {
  id: number;
  date: string;
  driverId: number;
  conductorId: number | null;
  busId: number;
  voucherNumber: number;
  commission: number | null;
  diesel: number | null;
  dieselLitres: number | null;
  cOilTechnician: number | null;
  toll: number | null;
  cleaning: number | null;
  alliedmor: number | null;
  cityParchi: number | null;
  refreshment: number | null;
  revenue: number | null;
  generator:number | null;
  miscellaneousExpense:number | null;
  repair:number | null;
  explanation:string;
  routeId?: number;
  isSubmitted?: boolean;
};

// Define the initial state with BusClosingVoucher type
interface BusClosingVoucherState {
  busClosingVouchers: BusClosingVoucher[];
}

// Helper function to generate current ISO DateTime for Date field
const getCurrentISODateTime = (): string => {
  return new Date().toISOString().split('T')[0];
};

const initialState: BusClosingVoucherState = {
  busClosingVouchers: []
};

// Create the slice
const busClosingVoucherSlice = createSlice({
  name: 'busClosingVouchers',
  initialState,
  reducers: {
    setBusClosingVoucher: (
      state,
      action: PayloadAction<Omit<BusClosingVoucher[], 'id'>>
    ) => {
      state.busClosingVouchers = action.payload;
    },
    // Action to add a new BusClosingVoucher
    // Action to add a new BusClosingVoucher
    addBusClosingVoucher: (
      state,
      action: PayloadAction<Omit<BusClosingVoucher, 'id'>>
    ) => {
      const newId =
        state.busClosingVouchers.length > 0
          ? state.busClosingVouchers[state.busClosingVouchers.length - 1].id + 1
          : 1;

      const newBusClosingVoucher: BusClosingVoucher = {
        ...action.payload,
        id: newId,
        date: getCurrentISODateTime()
      };

      // Push the new voucher to the state (mutating the state directly)
      state.busClosingVouchers.push(newBusClosingVoucher);
    },

    // Action to remove a BusClosingVoucher by ID
    removeBusClosingVoucher: (state, action: PayloadAction<number>) => {
      state.busClosingVouchers = state.busClosingVouchers.filter(
        (voucher) => voucher.id !== action.payload
      );
    },

    // Action to update an existing BusClosingVoucher by ID
    updateBusClosingVoucher: (
      state,
      action: PayloadAction<BusClosingVoucher>
    ) => {
      const index = state.busClosingVouchers.findIndex(
        (voucher) => voucher.id === action.payload.id
      );
      if (index !== -1) {
        state.busClosingVouchers[index] = action.payload;
      }
    }
  }
});

// Export actions and reducer
export const {
  addBusClosingVoucher,
  removeBusClosingVoucher,
  updateBusClosingVoucher,
  setBusClosingVoucher
} = busClosingVoucherSlice.actions;
export default busClosingVoucherSlice.reducer;

// Selector to get all bus closing vouchers
export const allBusClosingVouchers = (state: RootState) =>
  state.busClosingVouchers.busClosingVouchers;

File: BusServiceAPI/BusServiceAPI/Models/BusClosingVoucher.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace BusServiceAPI.Models
{
    public class BusClosingVoucher
    {
        public int Id { get; set; } // Primary key, auto-incremented
        public bool IsSubmitted { get; set; } = false;
        public DateTime? Date { get; set; }
        public int? VoucherNumber { get; set; }
        public int? Commission { get; set; }
        public int? Diesel { get; set; }
        public double? DieselLitres { get; set; }
        public int? COilTechnician { get; set; }
        public int? Toll { get; set; }
        public int? Cleaning { get; set; }
        public int? Alliedmor { get; set; }
        public int? CityParchi { get; set; }
        public int? Refreshment { get; set; }

        [Column(TypeName = "decimal(18,2)")]
        public decimal? Revenue { get; set; }
        public int? Repair { get; set; }
        public int? Generator { get; set; }
        public int? MiscellaneousExpense { get; set; }
        public string? Explanation { get; set; }

        // Navigation properties
        public Bus? Bus { get; set; }
        public Employee? Driver { get; set; }
        public Employee? Conductor { get; set; }
        public int? BusId { get; set; }
        public int? DriverId { get; set; }
        public int? ConductorId { get; set; }
        public Route? Route { get; set; }
        public int? RouteId { get; set; }
    }

}
